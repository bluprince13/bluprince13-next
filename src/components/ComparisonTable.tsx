import * as React from 'react'
import {
    Table,
    TableBody,
    TableCell,
    TableContainer,
    TableHead,
    TableRow,
    Paper,
    Typography,
    Divider,
    Box
} from '@mui/material'
import Markdown from 'react-markdown'

interface CellContent {
    content: string | React.ReactNode
    status: 'bad' | 'neutral' | 'good'
}

interface Header {
    name: string
    description?: string
}

interface TableProps {
    columnHeaders: Header[]
    rowHeaders: Header[]
    rows: CellContent[][]
    summary: string
}

const getStatusColor = (status: 'bad' | 'neutral' | 'good') => {
    switch (status) {
        case 'bad':
            return '#FFEBEE'
        case 'neutral':
            return '#FFFDE7'
        case 'good':
            return '#E8F5E9'
        default:
            return 'white'
    }
}

const tableCellSx = {
    backgroundColor: 'background.paper',
    border: '1px solid rgba(224, 224, 224, 1)',
    '&:hover': {
        backgroundColor: 'grey.200'
    }
}

const TableHeaderCell: React.FC<{
    name: string
    description?: string
}> = ({ name, description }) => {
    return (
        <TableCell sx={tableCellSx}>
            <Typography
                variant="subtitle1"
                sx={{ fontWeight: 'bold', textAlign: 'center' }}
            >
                <Markdown>{name}</Markdown>
            </Typography>
            {description && (
                <Typography
                    variant="body2"
                    sx={{
                        fontSize: '0.8rem',
                        textAlign: 'center',
                        maxWidth: 200,
                        wordWrap: 'break-word',
                        margin: 'auto'
                    }}
                >
                    <Markdown>{description}</Markdown>
                </Typography>
            )}
        </TableCell>
    )
}

export const ComparisonTable: React.FC<TableProps> = ({
    columnHeaders,
    rowHeaders,
    rows,
    summary
}) => {
    return (
        <Box>
            <TableContainer component={Paper}>
                <Table sx={{ minWidth: 650 }} aria-label="comparison table">
                    <TableHead>
                        <TableRow>
                            <TableCell sx={tableCellSx}></TableCell>
                            {columnHeaders.map(({ name, description }) => (
                                <TableHeaderCell
                                    key={name}
                                    name={name}
                                    description={description}
                                />
                            ))}
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {rows.map((row, rowIndex) => (
                            <TableRow key={rowIndex}>
                                <TableHeaderCell
                                    key={rowHeaders[rowIndex].name}
                                    name={rowHeaders[rowIndex].name}
                                    description={
                                        rowHeaders[rowIndex].description
                                    }
                                />
                                {row.map((cell, cellIndex) => (
                                    <TableCell
                                        key={cellIndex}
                                        align="center"
                                        sx={{
                                            ...tableCellSx,
                                            backgroundColor: getStatusColor(
                                                cell.status
                                            )
                                        }}
                                    >
                                        <span>{cell.content}</span>
                                    </TableCell>
                                ))}
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
            {summary && (
                <>
                    <Typography
                        variant="body1"
                        sx={{ fontWeight: 'bold', marginTop: 2 }}
                    >
                        Summary:
                    </Typography>
                    <Typography variant="body1">{summary}</Typography>
                </>
            )}
            <Divider sx={{ marginTop: 2, marginBottom: 4 }} />
        </Box>
    )
}
