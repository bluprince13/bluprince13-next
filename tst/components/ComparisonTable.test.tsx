import { render } from '@testing-library/react'
import { ComparisonTable } from '@Components/ComparisonTable'

jest.mock('react-markdown', () => {
    return ({ children }) => {
        return children
    }
})

describe('ComparisonTable', () => {
    const setup = ({} = {}) => {
        const queries = render(
            <ComparisonTable
                columnHeaders={[
                    {
                        name: 'Feature A',
                        description: 'Description for Feature A'
                    },
                    {
                        name: 'Feature B',
                        description: 'Description for Feature B'
                    },
                    {
                        name: 'Feature C',
                        description: 'Description for Feature C'
                    }
                ]}
                rowHeaders={[
                    {
                        name: 'Option 1',
                        description: 'Description for Option 1'
                    },
                    {
                        name: 'Option 2',
                        description: 'Description for Option 2'
                    },
                    {
                        name: 'Option 3',
                        description: `A very long descripition that needs more than one line`
                    }
                ]}
                rows={[
                    [
                        {
                            content: 'Good',
                            status: 'good'
                        },
                        {
                            content: 'Average',
                            status: 'neutral'
                        },
                        {
                            content: 'Poor',
                            status: 'bad'
                        }
                    ],
                    [
                        {
                            content: 'Poor',
                            status: 'bad'
                        },
                        {
                            content: 'Good',
                            status: 'good'
                        },
                        {
                            content: 'Average',
                            status: 'neutral'
                        }
                    ],
                    [
                        {
                            content: 'Average',
                            status: 'neutral'
                        },
                        {
                            content: 'Poor',
                            status: 'bad'
                        },
                        {
                            content:
                                'Very long content that needs more than one line',
                            status: 'good'
                        }
                    ]
                ]}
                summary="summary"
            />
        )
        return queries
    }

    it('renders correctly', () => {
        setup()

        const { asFragment } = setup()

        expect(asFragment()).toMatchSnapshot()
    })
})
