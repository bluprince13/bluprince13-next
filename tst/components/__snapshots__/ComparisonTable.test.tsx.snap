// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComparisonTable renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="comparison table"
        class="MuiTable-root css-quj9j6-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ywhbpj-MuiTableCell-root"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ywhbpj-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1qtbcgy-MuiTypography-root"
              >
                Feature A
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-l3td95-MuiTypography-root"
              >
                Description for Feature A
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ywhbpj-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1qtbcgy-MuiTypography-root"
              >
                Feature B
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-l3td95-MuiTypography-root"
              >
                Description for Feature B
              </p>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ywhbpj-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1qtbcgy-MuiTypography-root"
              >
                Feature C
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-l3td95-MuiTypography-root"
              >
                Description for Feature C
              </p>
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3dnjc9-MuiTableCell-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1qtbcgy-MuiTypography-root"
              >
                Option 1
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-l3td95-MuiTypography-root"
              >
                Description for Option 1
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-10396rc-MuiTableCell-root"
            >
              <span>
                Good
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-bgr10x-MuiTableCell-root"
            >
              <span>
                Average
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-5tm5lg-MuiTableCell-root"
            >
              <span>
                Poor
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3dnjc9-MuiTableCell-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1qtbcgy-MuiTypography-root"
              >
                Option 2
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-l3td95-MuiTypography-root"
              >
                Description for Option 2
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-5tm5lg-MuiTableCell-root"
            >
              <span>
                Poor
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-10396rc-MuiTableCell-root"
            >
              <span>
                Good
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-bgr10x-MuiTableCell-root"
            >
              <span>
                Average
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-3dnjc9-MuiTableCell-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1qtbcgy-MuiTypography-root"
              >
                Option 3
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body2 css-l3td95-MuiTypography-root"
              >
                A very long descripition that needs more than one line
              </p>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-bgr10x-MuiTableCell-root"
            >
              <span>
                Average
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-5tm5lg-MuiTableCell-root"
            >
              <span>
                Poor
              </span>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-10396rc-MuiTableCell-root"
            >
              <span>
                Very long content that needs more than one line
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1cl54jn-MuiTypography-root"
    >
      Summary:
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      summary
    </p>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-mm3ws1-MuiDivider-root"
    />
  </div>
</DocumentFragment>
`;
